x/emissions/abci.go

```diff
@@ -185,31 +185,48 @@ func distributeRewardsForMaturedBallots(
 			continue
 		}
 		ballots = append(ballots, ballot)
-		totalRewardsUnits += ballot.BuildRewardsDistribution(rewardsDistributeMap)
+		// Definitions:
+		// Correct votes: Votes which are in line with the final status of the ballot
+		// Incorrect votes: Votes which are not in line with the final status of the ballot
+		// Finalized Ballot: A ballot which has been finalized and is not in progress.Both success and failure ballots are finalized.
+		// Net Positive: Correct and Incorrect votes cancel each other out.
+
+		// Process votes and update rewardsMap
+		// Observer rewards are as follows:
+		// 1. Rewarded for correct votes
+		// 2. Penalized for incorrect votes
+
+		// After calculating the rewardsMap, the net positive observers are rewarded.
+		ballot.BuildRewardsDistribution(rewardsDistributeMap)
 	}
+	sortedKeys := make([]string, 0, len(rewardsDistributeMap))
+
+	for address, rewardUnits := range rewardsDistributeMap {
+		sortedKeys = append(sortedKeys, address)
+		// Rewards are only distributed for correct votes,calculate the total rewards units from the final map to allocate maximum rewards possible to observers
+		if rewardUnits > 0 {
```

Comment:
Not needed

Reply:
You mean `rewardUnits > 0`?

This is needed, as we only allocate rewards for correct votes. This makes sure the entire rewards are utilizied

Reply:
I moved the totalRewards caluculation out , so this should not change the allocation .
Just the portion of rewards which was allocated for incorrect observers would not now be distributed among others instreat of being retained by protocol